### PR TITLE
feat(extensions): add Enable HTTPS extension (Let's Encrypt / certbot DNS-01)

### PR DESCRIPTION
**Pull request by lexfrei**
## What

Adds an optional "Enable HTTPS" extension. It obtains a Let's Encrypt certificate via certbot's DNS-01 challenge and reconfigures the Mainsail/Fluidd nginx site into an HTTP→HTTPS redirect plus a TLS server block on port 443.

DNS-01 is used instead of HTTP-01 so it works on a printer that is only reachable on the LAN: HTTP-01 needs the CA to reach `http://<fqdn>/.well-known/...` from the public internet, which a private-IP host cannot satisfy. DNS-01 only needs an API token for a domain whose DNS is hosted by a supported provider. Supported here: Cloudflare, DigitalOcean and Linode; the provider model is a small dataclass, so adding more is a single entry.

## Why

KIAUH installs Mainsail/Fluidd behind nginx on port 80, HTTP only, with no built-in TLS option. Every LAN-only user currently hand-rolls a certbot DNS-01 issuance plus an nginx rewrite by hand. This turns it into a guided, reversible action.

## Behaviour

- Enable: gate on a Debian-based host, pick the client and DNS provider, prompt for the FQDN, email and credentials, issue the certificate, install a renewal reload hook, then apply the nginx rewrite transactionally — validate with `nginx -t` and restore the previous config if it fails, so nginx is never left unable to load.
- Disable: regenerate the original plain-HTTP site from KIAUH's own template (restoring the original listen port, read back from the live config) and reload. The certificate and credentials are kept; the manual cleanup commands are printed, not run.
- The provider token is written only to `/root/.secrets/<provider>.ini` (chmod 600 in a 700 dir), passed to the file via stdin, never as a command argument, never logged, never in KIAUH config. The prompt does not echo.

## Also included

- Fix `get_nginx_listen_port`: it took the last token of a `listen` line, so `listen 443 ssl http2;` parsed to `http2` and the port dropped to `None`. It now reads the address token after `listen`, so ssl / IPv6 / `host:port` forms resolve to the numeric port. Independent of the feature; happy to split it out.
- "Reconfigure Listen Port" now refuses to run on an HTTPS site (it would otherwise strip the TLS block and orphan the certificate).
- A no-echo secret-input helper (`get_secret_input`, via `getpass`) for credentials.

## Notes

- Maintainership: I'm happy to own this extension and act as its code owner going forward. For now `maintained_by` is set to `dw-0` by analogy with the other bundled extensions — glad to switch it to my username if you'd prefer I maintain it.
- Moonraker `cors_domains` is left unchanged: nginx is a same-origin reverse proxy, so the normal UI flow is not subject to CORS. The success dialog mentions the manual step only for the cross-origin case.
- Route53 / Google DNS are deferred — they use a different auth model (env/IAM, JSON key) rather than a single-token credentials file.

## Testing

86 new tests: pure transform with a golden snapshot, provider command/credentials assembly, secret handling, transactional apply + rollback, port round-trip, distro gate, listen-port parser edge cases, and the port-reconfigure guard. Full suite passes; ruff and mypy clean on the new modules.

### CODE REVIEW

File: kiauh/extensions/https/providers.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    without a credentials file composes without special-casing. All flags are
+    long form on purpose; the secret never appears here - only the path to the
+    credentials file does.
+    """
+    cmd: List[str] = ["sudo", "certbot", "certonly", provider.plugin_flag]
+    if provider.credentials_flag and credentials_path:
+        cmd += [provider.credentials_flag, credentials_path]
+    if provider.propagation_flag:
+        cmd += [provider.propagation_flag, str(propagation_seconds)]
+    cmd += [
```
Comment (dw-0) [inline kiauh/extensions/https/providers.py:134]: Do we need the `--no-eff-email` flag here as well? Referring to the docs of certbot, the default would be to ask if you want to share or not share your email with EFF: https://eff-certbot.readthedocs.io/en/latest/man/certbot.html#options

Since we want a fully interactive solution, we presumably don't want that question asked?

File: kiauh/extensions/https/nginx_https.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+_LISTEN_PORT_RE = re.compile(r"^\s*listen\s+(\d+)\b", re.MULTILINE)
+
+
+class AlreadyHttpsError(Exception):
+    """Raised when the given config already contains a TLS server block."""
+
+
+def is_https_config(config_text: str) -> bool:
+    """Return True if the config already contains a ``listen 443 ssl`` block."""
+    return _ALREADY_HTTPS_RE.search(config_text) is not None
```
Comment (dw-0) [inline kiauh/extensions/https/nginx_https.py]: I think if the input here would start with a `#` (so a commented line), the function would return true? You may want to make this a little bit more robust.

Maybe you want to extract this into a shared utils package, so you can use it across your extension and in the client_utils.py?

File: kiauh/components/webui_client/client_utils.py [hunk 3]
```diff
@@ -407,6 +410,18 @@ def get_nginx_listen_port(config: Path) -> int | None:
...
 
 
+def is_https_nginx_config(config: Path) -> bool:
+    """
+    Check whether an NGINX site already contains a TLS (HTTPS) server block.
+    :param config: The NGINX config file to inspect
+    :return: True if a 'listen 443 ssl' directive is present, else False
+    """
+    try:
+        return "listen 443 ssl" in config.read_text()
```
Comment (dw-0) [inline kiauh/components/webui_client/client_utils.py]: This needs to be more robust. Similar issue like https://github.com/dw-0/kiauh/pull/806/changes#r3487805965

File: kiauh/extensions/https/nginx_https.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    """
+    if is_https_config(original_config_text):
+        raise AlreadyHttpsError("Config already contains a 'listen 443 ssl' block.")
+
+    body = strip_listen_directives(extract_server_body(original_config_text))
+
+    redirect_block = (
+        "server {\n"
+        f"    listen {redirect_port};\n"
+        f"    listen [::]:{redirect_port};\n"
```
Comment (dw-0) [inline kiauh/extensions/https/nginx_https.py]: Will this cause any issues on systems where IPv6 is disabled?

File: kiauh/extensions/https/https_extension.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+            )
+            return None
+
+    def _apply_https(self, site: Path, new_config: str) -> bool:
+        """
+        Apply the rewritten config transactionally: back up, write, validate,
+        then reload - or restore the backup if nginx -t fails. The backup is
+        discarded either way. Returns True only when HTTPS is live and valid.
+        """
+        backup = backup_site(site)
```
Comment (dw-0) [inline kiauh/extensions/https/https_extension.py]: We may need a wrapping try except here. Several calls in here run a subprocess with check=True which can throw a CalledProcessError. That error is never handled.

File: kiauh/extensions/https/https_extension.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        fqdn = get_string_input(
+            "Printer FQDN (e.g. printer.example.com)", regex=FQDN_REGEX
+        )
+
+        # Build the rewritten config now, before any package install or
+        # certificate issuance. A site we cannot rewrite (malformed, no server
+        # block) must fail here - never after a certificate has been issued.
+        # The redirect block keeps the port the site is currently served on, so
+        # enabling and later disabling HTTPS round-trips to the same port.
+        redirect_port = get_nginx_listen_port(site) or HTTP_DEFAULT_PORT
```
Comment (dw-0) [inline kiauh/extensions/https/https_extension.py]: Do we really want to fallback to the default port here? A fallback to port 80, although another customized port is required, would lead to issues further down the road. We may want to abort here if we cannot get the port from the nginx config?

**Review by dw-0 — COMMENTED** [review]
Hi, thank you for your contribution!
I would like you to maintain this extension, yes. So please go ahead and write yourself into the metadata.json.

I am not very firm with the whole certbot thing, and i did a first rough review of the code. I made some comments we may want to clarify.

File: kiauh/extensions/https/providers.py (hunk 1 above)
Comment (lexfrei) [reply kiauh/extensions/https/providers.py:134]: Good catch — added `--no-eff-email`. Under `--non-interactive` certbot can't prompt, so this makes the EFF opt-out explicit rather than leaving it to certbot's default.

File: kiauh/extensions/https/nginx_https.py (hunk 2 above)
Comment (lexfrei) [reply kiauh/extensions/https/nginx_https.py]: Done both. Detection is now a single comment-aware `config_enables_https` in `utils/nginx_utils.py` — it strips `#` comments before matching and handles parameter ordering (`listen 443 ssl http2`, `listen [::]:443 ssl`). Both the extension and `client_utils` use it.

File: kiauh/components/webui_client/client_utils.py (hunk 3 above)
Comment (lexfrei) [reply kiauh/components/webui_client/client_utils.py]: `is_https_nginx_config` now delegates to the shared `config_enables_https` from the comment above, so a commented-out `# listen 443 ssl` is no longer a false positive here either.

File: kiauh/extensions/https/nginx_https.py (hunk 4 above)
Comment (lexfrei) [reply kiauh/extensions/https/nginx_https.py]: Right — an active `listen [::]` aborts `nginx -t` where IPv6 is off. Both generated blocks now emit the IPv6 listen commented out, matching KIAUH's own site template.

File: kiauh/extensions/https/https_extension.py (hunk 5 above)
Comment (lexfrei) [reply kiauh/extensions/https/https_extension.py]: Wrapped. The transactional apply catches `CalledProcessError` from every privileged step, restores this run's backup, and surfaces a clean error; the backup is kept if the restore itself fails.

File: kiauh/extensions/https/https_extension.py (hunk 6 above)
Comment (lexfrei) [reply kiauh/extensions/https/https_extension.py]: Agreed — guessing 80 would break the enable/disable round-trip. It now aborts when the listen port can't be read (and when it reads back as 443 for a site set up outside this extension).

**Comment by lexfrei** [comment]
Thanks — addressed all the inline comments (replies on each) and set `maintained_by` to me in `metadata.json` as you suggested.

One dependency worth flagging: the extension relies on `get_existing_clients()` to find the Mainsail/Fluidd install, and that helper currently returns an empty list on `develop` (it calls `typing.get_args()` on an `Enum`). I opened #812 with the one-line fix and tests; this PR needs it to work end-to-end.

I also added a `.github/CODEOWNERS` mapping each bundled extension to its author. Caveat: GitHub only auto-requests code owners with write access, so the external-author lines (mine included) are documentation until access is granted. Since I'll be maintaining the HTTPS extension, would you grant me access so its entry routes reviews to me? Happy to trim the file to just the HTTPS line if you'd rather not introduce repo-wide ownership in this PR.

**Comment by The-Monkey-King** [comment]
I just tried to run the changes outside of test utilities in a dev environment (just like a regular install) and I got an error on starting KIAUH.

`ImportError: cannot import name 'set_listen_port' from 'components.webui_client.client_utils' (/home/print1/kiauh/kiauh/components/webui_client/client_utils.py)
`

I know just enough to be extremely dangerous so a goof on my part may be to blame.  


```
Traceback (most recent call last):
  File "/home/print1/kiauh/kiauh/main.py", line 13, in <module>
    from core.menus.main_menu import MainMenu
  File "/home/print1/kiauh/kiauh/core/menus/main_menu.py", line 31, in <module>
    from core.menus.install_menu import InstallMenu
  File "/home/print1/kiauh/kiauh/core/menus/install_menu.py", line 24, in <module>
    from components.webui_client.menus.client_install_menu import ClientInstallMenu
  File "/home/print1/kiauh/kiauh/components/webui_client/menus/client_install_menu.py", line 16, in <module>
    from components.webui_client.client_utils import (
    ...<4 lines>...
    )
```

**Comment by The-Monkey-King** [comment]
> I just tried to run the changes outside of test utilities in a dev environment (just like a regular install) and I got an error on starting KIAUH.
> 
> `ImportError: cannot import name 'set_listen_port' from 'components.webui_client.client_utils' (/home/print1/kiauh/kiauh/components/webui_client/client_utils.py) `
> 
> I know just enough to be extremely dangerous so a goof on my part may be to blame.
> 
> ```
> Traceback (most recent call last):
>   File "/home/print1/kiauh/kiauh/main.py", line 13, in <module>
>     from core.menus.main_menu import MainMenu
>   File "/home/print1/kiauh/kiauh/core/menus/main_menu.py", line 31, in <module>
>     from core.menus.install_menu import InstallMenu
>   File "/home/print1/kiauh/kiauh/core/menus/install_menu.py", line 24, in <module>
>     from components.webui_client.menus.client_install_menu import ClientInstallMenu
>   File "/home/print1/kiauh/kiauh/components/webui_client/menus/client_install_menu.py", line 16, in <module>
>     from components.webui_client.client_utils import (
>     ...<4 lines>...
>     )
> ```

I do have the nginx_utils.py file in the utils directory. And listen_ports is defined. Here is the definition section as copied from github:

```
def listen_ports(config_text: str) -> set[int]:
    """
    Return the set of ports of every ACTIVE (non-commented) ``listen``
    directive in the config, across all server blocks.

    The port is taken from the address token right after ``listen``, keeping the
    part after the last ``:`` so ``listen 80;``, ``listen [::]:443 ssl;`` and
    ``listen 443 ssl http2;`` all yield the numeric port. Unlike scanning only
    the last listen per file, this sees a 443 directive that is followed by
    another listen - and unlike the ssl-aware check it also catches a plain,
    non-TLS ``listen 443;`` that would still collide on the socket.

    :param config_text: the full nginx config text
    :return: the set of active listen ports
    """
    ports: set[int] = set()
    for raw_line in config_text.splitlines():
        code = raw_line.split("#", 1)[0].strip()
        if not code.startswith("listen"):
            continue
        fields = code[len("listen") :].strip().rstrip(";").split()
        if not fields:
            continue
        token = fields[0].split(":")[-1].strip("[]")
        if token.isdigit():
            ports.add(int(token))
    return ports
```

**Comment by lexfrei** [comment]
@The-Monkey-King I can't reproduce it. A fresh clone of the branch at its current HEAD starts cleanly for me both on a real Sovol Zero (Debian Bullseye, Python 3.9) with Mainsail installed and in Debian containers (Python 3.11 and 3.13) — the main menu loads and `set_listen_port` imports fine. (`set_listen_port` is existing KIAUH code, not something this PR adds, so any current checkout has it.)

Since you also have nginx_utils/listen_ports, this looks like a mixed checkout rather than a missing file: the branch was rebased and force-pushed, and `git pull` won't fast-forward across that, so a local copy can end up with files at inconsistent revisions. A clean re-sync should sort it:

```
git fetch origin
git reset --hard origin/feat/enable-https-certbot-extension   # discards local edits
```

If it still fails after a clean checkout, could you share your OS, `python3 --version`, and how you pulled the branch? I'll reproduce your exact setup from that.

Heads-up once it starts: enabling HTTPS calls `get_existing_clients()`, which returns `[]` on `develop` (it calls `typing.get_args()` on an `Enum`) — #812 has the fix; without it you'll get "No Mainsail or Fluidd installation was found".

**Comment by lexfrei** [comment]
I also verified this end-to-end on real hardware, not just the mocked unit tests.

On a Debian Bullseye host (Python 3.9) with a KIAUH Mainsail install and an existing Let's Encrypt cert (Cloudflare DNS-01), I ran the extension's install (enable) and remove (disable) actions against the live nginx and certbot:

- Disable reverse-transformed the running HTTPS site back to plain HTTP, preserving the proxy/location directives, on the original port.
- Re-enable reused the saved Cloudflare credentials, certbot kept the still-valid cert (`--keep-until-expiring` — no re-issue, so no Let's Encrypt rate-limit), installed the renew hook, and applied the TLS config transactionally (`nginx -t` + reload). `https://` → 200, `http://` → 301.
- Restoring my original site file afterwards produced a byte-identical result.

This exercised the paths the unit tests mock: the certbot invocation, the `sudo` credential/nginx writes, and `nginx -t` + reload.

### Repro

On a Debian host with a KIAUH Mainsail install and a domain whose DNS is on Cloudflare / DigitalOcean / Linode:

1. Check out this branch **plus #812** — `get_existing_clients()` returns `[]` on `develop` (it calls `typing.get_args()` on an `Enum`), so without that fix the extension reports "No Mainsail or Fluidd installation was found".
2. `./kiauh.sh` → `E) Extensions` → the HTTPS extension → Install → pick the DNS provider, enter the FQDN + contact email, paste a scoped DNS-edit token, confirm. certbot runs the DNS-01 challenge and the Mainsail site becomes an `:80`→`:443` redirect plus a TLS block; open `https://<fqdn>`.
3. Remove via the same menu → the site reverts to plain HTTP (custom directives kept) on its original port.

If the FQDN already has a valid certificate, re-enabling reuses the saved credentials and certbot keeps the existing cert instead of re-issuing.

**Comment by The-Monkey-King** [comment]
Thanks. It was a one-off for me too. I wiped the container and started fresh. 
 By the way, my configuration is:
- Windows/WSL
- Debian 13.5 (from Debian website)
- Installed python 3.13.5 and git
- cloned kiauh as https://github.com/lexfrei/kiauh.git
- re-synched and reset 

NOTE: I am not connected to a printer at this point. I am only testing the MCU and not a client. If it needs a client then I'll fetch one. 

Menu runs. 
 - I declined to update kiauh. 
 - Installed Klipper, then Moonraker, then Fluidd
 - Entered (E) Extensions and the menu throws error that 'Enable HTTPS' is dropped because the menu #(15) is in conflict with 'DroidKlipp'
 - No problem. I exit out of kiauh and set Enable HTTPS menu to 17 (first open menu number). 
 - Try to install but it says Mainsail or Fluidd is not installed. 
 - Removed Fluidd and install Mainsail - same issue

All in all, I am just chalking this up to a one-off issue with my setup as a corner case. I can get out to the Internet with my WSL distro so I don't think it is a connectivity issue. 

So please don't waste any time with me. I was interested in your method of installing HTTPS as I've already seen instructions for OpenSSL. And again, sorry for introducing a corner case. 



> @The-Monkey-King I can't reproduce it. A fresh clone of the branch at its current HEAD starts cleanly for me both on a real Sovol Zero (Debian Bullseye, Python 3.9) with Mainsail installed and in Debian containers (Python 3.11 and 3.13) — the main menu loads and `set_listen_port` imports fine. (`set_listen_port` is existing KIAUH code, not something this PR adds, so any current checkout has it.)
> 
> Since you also have nginx_utils/listen_ports, this looks like a mixed checkout rather than a missing file: the branch was rebased and force-pushed, and `git pull` won't fast-forward across that, so a local copy can end up with files at inconsistent revisions. A clean re-sync should sort it:
> 
> ```
> git fetch origin
> git reset --hard origin/feat/enable-https-certbot-extension   # discards local edits
> ```
> 
> If it still fails after a clean checkout, could you share your OS, `python3 --version`, and how you pulled the branch? I'll reproduce your exact setup from that.
> 
> Heads-up once it starts: enabling HTTPS calls `get_existing_clients()`, which returns `[]` on `develop` (it calls `typing.get_args()` on an `Enum`) — #812 has the fix; without it you'll get "No Mainsail or Fluidd installation was found".

**Comment by lexfrei** [comment]
The dependency I flagged earlier is stale: `develop` fixed `get_existing_clients()` itself in 09dd829, so this PR no longer depends on anything and I've closed #812. Rebased onto current `develop`; CI is green on 3.8 and 3.11.

**Comment by lexfrei** [comment]
@dw-0 friendly ping on this one. All the inline feedback from your review is addressed, the branch is rebased onto current `develop`, and CI is green on 3.8 and 3.11. It's also been verified end-to-end on real hardware, not just the mocked unit tests. Happy to rework anything if the approach or scope isn't what you'd want for a bundled extension — just say the word.
